### ISSUE TITLE
archive-async issues

### ISSUE THREAD

**Issue by bruno-** [pgbackrest/pgbackrest#1995]
First off: thank you for maintaining this fantastic tool. I'm very grateful for all the time you dedicate to it 🙏

1. pgBackRest version:

pgBackRest 2.44

2. PostgreSQL version:

```bash
$ /usr/lib/postgresql/15/bin/postgres --version
postgres (PostgreSQL) 15.2 (Ubuntu 15.2-1.pgdg22.04+1)
```

4. Operating system/version - if you have more than one server (for example, a database server, a repository host server, one or more standbys), please specify each:

All the servers use this:

```bash
$ lsb_release -a
No LSB modules are available.
Distributor ID: Ubuntu
Description:    Ubuntu 22.04.1 LTS
Release:        22.04
Codename:       jammy
```

6. Did you install pgBackRest from source or from a package?

Compiled from source.

7. Please attach the following as applicable:
    - `pgbackrest.conf` file(s)
      - [dedicated_repository_host_pgbackrest.conf.txt](https://github.com/pgbackrest/pgbackrest/files/10716192/dedicated_repository_host_pgbackrest.conf.txt)
      - [database_pgbackrest.conf.txt](https://github.com/pgbackrest/pgbackrest/files/10716193/database_pgbackrest.conf.txt)
    - `postgresql.conf` settings applicable to pgBackRest
      - `archive_command` - `pgbackrest --stanza=main archive-push %p`
      - `archive_mode` - `on`
      - `listen_addresses` - `10.0.0.5, 127.0.0.1`
      - `max_wal_senders` - not changed, the default (10)
      - `wal_level` - changed, the default (replica)
      - `port` - `5432`
      -  Full config: [postgresql.conf.txt](https://github.com/pgbackrest/pgbackrest/files/10716196/postgresql.conf.txt)
    - errors in the postgresql log file before or during the time you experienced the issue: attached below.
    - log file in `/var/log/pgbackrest` for the commands run (e.g. `/var/log/pgbackrest/mystanza_backup.log`)

8. Describe the issue:

TL;DR: `archive-push` and `archive-get` commands error when postgresql process runs them via `archive_command` and during recovery respectively. If `archive-async` is set to `n`, everything works flawlessly.

- I'm using an empty database with no load.
- I can provide ssh access to the maintainers to the servers that exhibit these problems.

**archive-push problems**

- I run `pgbackrest --stanza=main check` on the dedicated repository host.
- archive-push command fails on the postgresql side: [archive_push_postgresql_log.txt](https://github.com/pgbackrest/pgbackrest/files/10716205/archive_push_postgresql_log.txt)
- The log file `/var/log/pgbackrest/main-archive-push-async.log` is NOT created.
- If I run the failing command manually in the console, everything works flawlessly and the log file `/var/log/pgbackrest/main-archive-push-async.log` is created:

```bash
$ pgbackrest --stanza=main archive-push pg_wal/00000005000000000000002D
2023-02-12 13:52:15.033 P00   INFO: archive-push command begin 2.44: [pg_wal/00000005000000000000002D] --archive-async --exec-id=17715-a4687d2a --log-level-console=info --log-level-file=detail --pg1-path=/var/lib/postgresql/15/main --repo1-host=10.0.0.10 --stanza=main                                            
2023-02-12 13:52:17.040 P00   INFO: pushed WAL file '00000005000000000000002D' to the archive asynchronously
2023-02-12 13:52:17.040 P00   INFO: archive-push command end: completed successfully (2011ms)           
postgres@qa-01-main-database-01:~/15/main$ pwd                                                          
/var/lib/postgresql/15/main                          
```
- If I change `archive-async=n` on the database server and repeat the steps, there are no problems:
```bash
2023-02-12 14:16:59.410 UTC [18102] postgres@postgres LOG:  restore point "pgBackRest Archive Check" created at 0/300001B0
2023-02-12 14:16:59.410 UTC [18102] postgres@postgres STATEMENT:  select pg_catalog.pg_create_restore_point('pgBackRest Archive Check')::text
2023-02-12 14:16:59.567 P00   INFO: archive-push command begin 2.44: [pg_wal/000000050000000000000030] --no-archive-async --exec-id=18104-39219869 --log-level-console=info --log-level-file=detail --pg1-path=/var/lib/postgresql/15/main --repo1-host=10.0.0.10 --stanza=main
2023-02-12 14:17:00.715 P00   INFO: pushed WAL file '000000050000000000000030' to the archive
2023-02-12 14:17:00.822 P00   INFO: archive-push command end: completed successfully (1259ms)
```
The log file for the successful `archive-push` is still not created in `/var/log/pgbackrest`, not sure if it should?

**archive-get problems**

- `sudo systemctl stop postgresql`
- As `postgres` user: `pgbackrest --stanza=main restore  --delta`
```bash
2023-02-12 14:23:02.291 P00   INFO: restore command begin 2.44: --delta --exec-id=18159-957e0b9b --log-level-console=info --log-level-file=detail --pg1-path=/var/lib/postgresql/15/main --repo1-host=10.0.0.10 --stanza=main
2023-02-12 14:23:02.833 P00   INFO: repo1: restore backup set 20230212-063003F, recovery will start at 2023-02-12 06:30:03
2023-02-12 14:23:02.838 P00   INFO: remove invalid files/links/paths from '/var/lib/postgresql/15/main'
2023-02-12 14:23:03.966 P00   INFO: write updated /var/lib/postgresql/15/main/postgresql.auto.conf
2023-02-12 14:23:03.990 P00   INFO: restore global/pg_control (performed last to ensure aborted restores cannot be started)
2023-02-12 14:23:03.991 P00   INFO: restore size = 61.2MB, file total = 1261
2023-02-12 14:23:03.992 P00   INFO: restore command end: completed successfully (1706ms)
```
- `sudo systemctl start postgresql` - relevant log showing `archive-get` failures: [archive_get_postgresql_log.txt](https://github.com/pgbackrest/pgbackrest/files/10716272/archive_get_postgresql_log.txt).
- I manually checked and the files ARE present in the dedicated repository host archive.
- Log file `/var/log/pgbackrest/main-archive-get-async.log` is NOT created.
- If I manually run `archive-get` from the console it succeeds:
```bash
$ pgbackrest --stanza=main archive-get pg_wal/000000030000000000000027 /var/lib/postgresql/15/main/pg_wal/000000030000000000000027
2023-02-12 14:35:34.196 P00   INFO: archive-get command begin 2.44: [pg_wal/000000030000000000000027, /var/lib/postgresql/15/main/pg_wal/000000030000000000000027] --archive-async --exec-id=18490-137b795a --log-level-console=info --log-level-file=detail --pg1-path=/var/lib/postgresql/15/main --repo1-host=10.0.0.10 --stanza=main
2023-02-12 14:35:34.199 P00   INFO: found 000000030000000000000027 in the archive asynchronously
2023-02-12 14:35:34.205 P00   INFO: archive-get command end: completed successfully (13ms)
```

-  If I change `archive-async=n` on the database server and repeat the steps, there are no problems, relevant log: [archive_get_no_async_postgresql_log.txt](https://github.com/pgbackrest/pgbackrest/files/10716285/archive_get_no_async_postgresql_log.txt)
- The log file for the successful `archive-get` is still not created in `/var/log/pgbackrest`, not sure if it should?

**Comment by pgstef** [comment]
Hi,

Not sure what's going on with your setup but since the archiving is happening on the database host, `archive-async=y` should be added in the pgbackrest.conf on your db host, not on the repo host.

Archive-get and archive-push don't have their own log files, the output goes to PG logs. The async processes have their own log files, on the db host normally. Does the system user used by your PG server has enough rights on `/var/log/pgbackrest`?

**Comment by bruno-** [comment]
@pgstef to speed up debugging I added [your SSH key](https://github.com/pgstef.keys) to the servers where this is happening. Feel free to poke around, there's nothing sensitive on these servers. `user` user has passwordless `sudo`.

- database server: `ssh -J user@91.107.137.178 user@10.0.0.5`
- dedicated repository host: `ssh -J user@91.107.137.178 user@10.0.0.10`

Let me know if I can help in any other way.

**Comment by pgstef** [comment]
> Let me know if I can help in any other way.

I won't have the time to do that. Sorry.

**Comment by bruno-** [comment]
I hope my last proposal didn't come off as "bossy" - that was definitely not the intention.
As an open source maintainer myself, I'm super grateful for the time you spent on this issue 🙏, even if you can't dedicate more. Thanks for working on a great tool.

**Comment by vitabaks** [comment]
@bruno- Please try specifying `log-level-console= detail` (or debug) to see more details about the archiving using pgbackrest in the Postgresql log.

**Comment by bruno-** [comment]
Good idea. Here's a relevant chunk from the postgresql log: [postgresql-15-main.log](https://github.com/pgbackrest/pgbackrest/files/10732749/postgresql-15-main.log)

Based on the log it seems the issue is related to `/var/spool/pgbackrest` dir. Not sure what as `/var/spool/pgbackrest` owner is `postgres:postgres`. I even set dir permissions to `0777` and the issue still happens.

**Comment by pgstef** [comment]
> I hope my last proposal didn't come off as "bossy" - that was definitely not the intention. As an open source maintainer myself, I'm super grateful for the time you spent on this issue pray, even if you can't dedicate more. Thanks for working on a great tool.

No problem, It's just that I'm traveling for work this week and couldn't promise to find time to be able to connect to some remote servers to investigate. (I tried but it doesn't seem to work though : `user@91.107.137.178: Permission denied (publickey).
kex_exchange_identification: Connection closed by remote host`)

It would be surprising but does the spool path also exists on the repo host with accurate permissions ?
Did you try adding more processes (ex `process-max=2`) ?

Is there any profile difference between the user used by your PG server and the user you use to run the commands manually?

**Comment by bruno-** [comment]
> I tried but it doesn't seem to work though user@91.107.137.178: Permission denied

Thank you for trying it out. Since you mentioned you won't have the time I revoked the access. I have enabled it again - so if you have the time feel free to try it out (no stress if not).

> does the spool path also exists on the repo host with accurate permissions ?

Yes, the default `/var/spool/pgbackrest` dir also exists on the repo host. The owner is `pgbackrest:pgbackrest`, permissions `0750`. I have loosened the permissions to `0777`, but no success.

> Is there any profile difference between the user used by your PG server and the user you use to run the commands manually?

To my knowledge there is no difference. The default `postgres` user runs the database. I'm also using `sudo su - postgres` to run `pgbackrest` commands on the postgresql host.

**Comment by bruno-** [comment]
Hi,

@vitabaks is helping me troubleshoot this issue. He mentioned it would be useful to bring up the fact that archiving works when triggered manually from the console, here's the example:

```log
$ pgbackrest --stanza=main archive-get pg_wal/000000030000000000000027 /var/lib/postgresql/15/main/pg_wal/000000030000000000000027
2023-02-12 14:35:34.196 P00   INFO: archive-get command begin 2.44: [pg_wal/000000030000000000000027, /var/lib/postgresql/15/main/pg_wal/000000030000000000000027] --archive-async --exec-id=18490-137b795a --log-level-console=info --log-level-file=detail --pg1-path=/var/lib/postgresql/15/main --repo1-host=10.0.0.10 --stanza=main
2023-02-12 14:35:34.199 P00   INFO: found 000000030000000000000027 in the archive asynchronously
2023-02-12 14:35:34.205 P00   INFO: archive-get command end: completed successfully (13ms)
```

It's only when `postgresql` process starts archiving that things fail.

**Comment by dwsteele** [comment]
> He mentioned it would be useful to bring up the fact that archiving works when triggered manually

This means that something is different about the env when running from the postgres server, which can be tricky to debug. You'll need to look at what you have done that is different than a vanilla install. We've seen this before, it always turns out to be some unusual thing that was done in the env that did not work as expected.

**Comment by bruno-** [comment]
Hi @dwsteele
thank you for chiming in.

> This means that something is different about the env when running from the postgres server, which can be tricky to debug. 

This makes sense. I wanted to confirm that's really the case so I created new server and provisioned it from scratch manually.

TL;DR: unfortunately I'm observing the same behavior with vanilla postgresql install. `archive-push` command fails when run from `postgresql`, it succeeds when run manually.

Here's what was done:

**Database server**

- Create user `user`, setup ssh
- Postgresql installation: `sudo apt-get update`, `sudo apt-get install -y postgresql` , `version: 'postgres (PostgreSQL) 14.6 (Ubuntu 14.6-0ubuntu0.22.04.1)'`
- The only change to default `postgresql.conf`: `archive_mode` and `archive_command`.
- Pgbackrest install as described in debian user guide, compiled from tarball, setup log dir, config, spool dir, passwordless SSH login to database repo server.

**Comment by bruno-** [comment]
Any pointers what should I try next?

Maybe try different hosting provider? I'm currently using `hetzner`, but maybe they somehow modify `ubuntu` images and it impacts postgresql and pgbackrest ? (wild guess)

**Comment by vitabaks** [comment]
@bruno- Try another repository mode, for example s3 (it could be a [MinIO](https://min.io/)). \
Check if there will be the same error?

**Comment by bruno-** [comment]
> Try another repository mode, for example s3

Ok, I just tried using a S3 compatible storage Backblaze B2. I'm getting the same behavior:

- `archive-push` errors when ran from postgresql.
- The command succeeds when ran manually from the command line.

**Comment by vitabaks** [comment]
@bruno- Thanks for checking this out.

@dwsteele David Steeleit's strange that we can't make pgbackrest work in async mode. Perhaps there really is some kind of bug here. Or we are doing something wrong. I'd really like to figure this out.

**Comment by dwsteele** [comment]
Could be a bug, but I'm not see any evidence of that. The evidence is that archive-push async is never getting run at all. 

The documentation is a piece of code that runs all the commands you see in containers. Plus we have tons of unit and integration tests that exercise async archiving. Pretty sure this is environmental.

Maybe @bruno- is right and this has something to do with the OS images he is using? Also -- I recommend never building from source. Get packages from apt.postgresql.org.

**Comment by pgstef** [comment]
> Also -- I recommend never building from source. Get packages from apt.postgresql.org.

To be complete, I tested this morning on 2 new vms (using `debian/bullseye64` vagrant box), with PG 15 and pgBackRest packages from PGDG repos and it working just fine with the following config:

```ini
## PG host
[global]
repo1-host=bck1
repo1-host-user=pgbackrest
log-level-console=info
log-level-file=debug
delta=y
process-max=2
archive-async=y

[demo]
pg1-path=/var/lib/postgresql/15/main

## Repo host
[global]
repo1-path=/backup_space
repo1-retention-full=2
log-level-console=info
log-level-file=debug
start-fast=y
delta=y

[demo]
pg1-host=pg1
pg1-path=/var/lib/postgresql/15/main
```

Didn't do anything special except the `pgbackrest` system user creation on the repo host + ssh access and `chown` some default directories on the repo host.

**Comment by vitabaks** [comment]
Thanks @dwsteele @pgstef 

@bruno- please try install pgbackrest from a package instead of building from source. And also check the configuration with the example above.

**Comment by bruno-** [comment]
> Also -- I recommend never building from source. Get packages from apt.postgresql.org.

After installing `pgbackrest` from `apt.postgresql.org` things work fine.
This is how I feel -> 🤦‍♂️

Maintainers: thank you for the patient and helpful support 🙏

**Comment by dwsteele** [comment]
Glad you got it working. Seems like issues (different each time) with building from source have come up a lot recently (see #1988, #1992). Packages are always better when they are an option.

**Comment by bruno-** [comment]
Ok, after checking those other issues and trying things out, I discovered that setting a **full path** to self-compiled pgbackrest binary works.

So, this made it work: `archive_command = '/usr/local/bin/pgbackrest --stanza=main archive-push %p'` 🤷‍♂️
